Add the Microsoft GDK platform

RandomShaper opened the pull request:
**UPDATE:**
This PR is getting big, since it will add GDK as a targetable platform. I'll update the description later.

Based on #70315 and #73212.

---
🍀 This work has been financed and kindly donated to the Godot Engine project by [**W4 Games**](https://w4games.com). 🍀

RandomShaper committed:
[D3D12] Enable for CI builds

RandomShaper committed:
Avoid Python code format checks in third-party code

RandomShaper committed:
[D3D12] Add third-party code

RandomShaper committed:
[D3D12] Add first-party code

bruvzg commented on drivers/SCsub:
```diff
@@ -11,9 +11,7 @@ SConscript("windows/SCsub")
-SConscript("pulseaudio/SCsub")
-if env["platform"] == "windows":
-    SConscript("wasapi/SCsub")
+SConscript("wasapi/SCsub")
```
`SConscript("pulseaudio/SCsub")` got accidentally removed.

RandomShaper replied:
Fixed in upcoming push.

bruvzg commented on thirdparty/games_are_not_uwp.patch:
```diff
@@ -0,0 +1,57 @@
+   file->contents = (char *) MapViewOfFileFromApp (file->mapping, FILE_MAP_READ, 0, 0);
+ #else
+   file->contents = (char *) MapViewOfFile (file->mapping, FILE_MAP_READ, 0, 0, 0);
+diff --git a/thirdparty/libwebp/src/utils/thread_utils.c b/thirdparty/libwebp/src/utils/thread_utils.c
```
Probably should be separate patches for HarfBuzz and WebP.

RandomShaper replied:
WebP was not really needed. Removed in upcoming push.

bruvzg commented on thirdparty/games_are_not_uwp.patch:
```diff
@@ -0,0 +1,57 @@
+diff --git a/thirdparty/README.md b/thirdparty/README.md
+index dd5184a5ce..c4be92d171 100644
+--- a/thirdparty/README.md
++++ b/thirdparty/README.md
+@@ -276,6 +276,8 @@ Files extracted from upstream source:
+ - the `src` folder
+ - `AUTHORS`, `COPYING`, `THANKS`
+ 
++Applied the patch in `patches/games_are_not_uwp.patch` (upstream PR: https://github.com/harfbuzz/harfbuzz/pull/4109).
++
+ 
+ ## icu4c
+ 
```
This part should not be in the patch.

RandomShaper replied:
I messed up the patch file. Happens sometimes... In any case, they've already included my patch in 7.0.0, so I'm including a commit to move to that version in my next push.

bruvzg reviewed (commented):
ICU / HarfBuzz changes seems fine to me.

RandomShaper committed:
Add the Microsoft GDK platform

- Avoid Windows platform build script from mistaking GDK environment for pure Windows.
- Let glslang module detect UNIX/non-UNIX more accurately.
- Let WASPI be restricted by WASAPI_ENABLED alone.
- Make MS flavors of NetSocketPosix and IPUnix available to any Win32 platform.
- Prevent icu4c from assuming Win32 is necessarily Windows.
- Make WASAPI buildable with COM lite.
- Make DirectWrite an optional feature.
- Adapt `OS_Windows` to pure gaming platforms.

RandomShaper committed:
Update HarfBuzz to 7.0.0

akien-mga commented:
Not urgent, but I would suggest moving the harfbuzz 7.0.0 to a dedicated PR (and rebase this PR on top of the new one). I don't know if we'd merge harfbuzz 7.0.0 for 4.0 (that's @bruvzg's call) but it's likely to be mergeable before this PR either way.

bruvzg commented:
> upgrade to HarfBuzz 7.0.0

Seems like build config will need to be updated, there are new features in 7.0.0 and probably new files to include.

RandomShaper commented:
#73212.

RandomShaper committed:
Let VS solution name be overridden
